docker_cli/image.py

```diff
@@ -0,0 +1,430 @@
+import io
+import os
+from shlex import split
+from subprocess import PIPE, STDOUT, CalledProcessError, run
+from sys import stdin
+from typing import (Any, List, Optional, Sequence, Type, TypeVar, Union,
+                    overload)
+
+from .exceptions import CommandNotFoundOnImageError
+
+
+class Image:
+    """A class which holds a reference to a given docker image and provides
+    an interface by which to interact with that image.
+
+    -   Capable of building docker images from dockerfiles or dockerfile-
+        formatted strings via the build classmethods.
+    -   Capable of acquiring information about the given image (see property
+        methods below).
+    -   Capable of running commands on containers built from the image using
+        the run method.
+    -   All docker build and run methods on the class can recive a file-like
+        object, (io.TextIOBase or some subclass of it,) and they will write the
+        outputs from docker into the file to enable logging.
+
+    NOTE: This class does not work correctly if the image built does not have
+        bash installed.
```

Comment:
I think this note is only applicable to the `run()` method -- the other Image functionality should work OK without bash installed -- so let's move this to the docstring of `run()`.

Reply:
This might be a good place to use an [admonition](https://docutils.sourceforge.io/docs/ref/rst/directives.html#specific-admonitions) like this:

```suggestion
    .. note:: 
       This class does not work correctly if the image built does not have
       bash installed.
```

If you render html from the docstring, the note will then appear in a highlighted box like this: https://sphinx-rtd-theme.readthedocs.io/en/stable/demo/demo.html#admonitions

Actually, it might make more sense to use a "warning" admonition rather than a "note" in this case:

```suggestion
    .. warning:: 
       This class does not work correctly if the image built does not have
       bash installed.
```

Reply:
In the long run, it might be possible to relax this requirement that the image must have bash installed. For example, the `run()` method could use the default [`ENTRYPOINT`](https://docs.docker.com/engine/reference/builder/#entrypoint) of the image instead of bash, and we could add an optional argument to override the entrypoint.

But for now, I'm pretty happy to keep this restriction in place, and we can worry about whether this functionality could be more generic later on.